microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/core/java/validators/annotations/RangeExpression.java

```diff
@@ -0,0 +1,220 @@
+/*******************************************************************************
+* Copyright (c) 2021 Red Hat Inc. and others.
+*
+* This program and the accompanying materials are made available under the
+* terms of the Eclipse Public License v. 2.0 which is available at
+* http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+* which is available at https://www.apache.org/licenses/LICENSE-2.0.
+*
+* SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+*
+* Contributors:
+*     Red Hat Inc. - initial API and implementation
+*******************************************************************************/
+package org.eclipse.lsp4mp.jdt.internal.core.java.validators.annotations;
+
+import java.text.MessageFormat;
+
+import org.apache.commons.lang3.StringUtils;
+
+/**
+ * The range expression used to validate value from an attribute annotation.
+ * 
+ * The range is specified like OSGi version syntax:
+ * 
+ * <ul>
+ * <li>0 -> means >=0</li>
+ * <li>[0 -> means >=0</li>
+ * <li>(0 -> means >0</li>
+ * <li>[0,1] -> means >=0 && <=1</li>
+ * <li>(0,1] -> means >0 && <=1</li>
+ * <li>(0,1) -> means >0 && <1</li>
+ * <li>[0,1) -> means >=0 && <1</li>
+ * </ul>
+ * 
+ * @author Angelo ZERR
+ *
+ */
+public class RangeExpression {
+
+	private final Edge from;
+	private final Edge to;
+	private final String errorMessage;
+
+	public RangeExpression(Edge from, Edge to) {
+		this.from = from;
+		this.to = to;
+		this.errorMessage = createErrorMessage(from, to);
+	}
+
+	private static String createErrorMessage(Edge from, Edge to) {

```

Comment:
One thing maybe worth changing is when the error message says (for example) :

`The value '-1' must be between '0' inclusive and '1' inclusive.`

The inclusive/exclusive should probably be in round brackets. It's not really part of the sentence but more of an aside.

Reply:
Fixed. I decided to use ``` insteadof using simple quote `'` in the message to have a markdown style. Even if LSP client doesn't support markdown, there will have ```  which I think it's not annoying.